mountpoint-s3-crt/src/auth/imds_client.rs

```diff
@@ -0,0 +1,119 @@
+//! A client for retrieving ec2 instance metadata.
+
+use crate::auth::auth_library_init;
+use crate::common::allocator::Allocator;
+use crate::common::error::Error;
+use crate::io::channel_bootstrap::ClientBootstrap;
+use crate::io::retry_strategy::RetryStrategy;
+use crate::CrtError;
+use mountpoint_s3_crt_sys::{
+    aws_byte_buf, aws_imds_client, aws_imds_client_get_instance_type, aws_imds_client_new, aws_imds_client_options,
+    aws_imds_client_release,
+};
+use std::ptr::NonNull;
+
+/// A client for instance metadata query.
+#[derive(Debug)]
+pub struct ImdsClient {
+    /// A pointer to the underlying `aws_imds_client`
+    inner: NonNull<aws_imds_client>,
+}
+
+/// Configurations for creating a new [ImdsClient].
+#[derive(Debug, Default)]
+pub struct ImdsClientConfig {
+    /// The structure that can pass into the CRT's functions.
+    inner: aws_imds_client_options,
+
+    /// The [ClientBootstrap] to use to create connection to IMDS.
+    client_bootstrap: Option<ClientBootstrap>,
+
+    /// The [RetryStrategy] to use to reschedule failed requests to IMDS. This is reference counted,
+    /// so we only need to hold onto it until this [ImdsClientConfig] is consumed, at which point the
+    /// client will take ownership.
+    retry_strategy: Option<RetryStrategy>,
+}
+
+impl ImdsClientConfig {
+    /// Create a new [ImdsClientConfig] with default options.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Client bootstrap used for common staples such as event loop group, host resolver, etc.
+    pub fn client_bootstrap(&mut self, client_bootstrap: ClientBootstrap) -> &mut Self {
+        self.inner.bootstrap = client_bootstrap.inner.as_ptr();
+        self.client_bootstrap = Some(client_bootstrap);
+        self
+    }
+
+    /// Retry strategy used to reschedule failed requests
+    pub fn retry_strategy(&mut self, retry_strategy: RetryStrategy) -> &mut Self {
+        self.inner.retry_strategy = retry_strategy.inner.as_ptr();
+        self.retry_strategy = Some(retry_strategy);
+        self
+    }
+}
+
+impl ImdsClient {
+    /// Create a new [ImdsClient]
+    pub fn new(allocator: &Allocator, config: ImdsClientConfig) -> Result<Self, Error> {
+        auth_library_init(allocator);
+
+        // SAFETY: `allocator` lives as long as [ImdsClient]. On crt side, `aws_imds_client_new` manages to increase
+        // reference counts of boot_strap and retry_strategy.
+        let inner = unsafe { aws_imds_client_new(allocator.inner.as_ptr(), &config.inner).ok_or_last_error()? };
+        Ok(Self { inner })
+    }
+
+    /// Get the EC2 instance type.
+    pub fn get_instance_type<F>(&self, callback: F) -> Result<(), Error>
+    where
+        F: FnOnce(Result<String, Error>) + 'static,
+    {
+        let callback_closure = Box::new(callback);
+        let callback_wrapper = Box::new(ImdsClientGetResourceCallback(Box::new(callback_closure)));
```

Comment:
One too many `Box`es:
```suggestion
        let callback_wrapper = Box::new(ImdsClientGetResourceCallback(Box::new(callback)));
```